File: src/main/java/io/github/kingg22/api/vacunas/panama/configuration/security/SecurityConfig.java
package io.github.kingg22.api.vacunas.panama.configuration.security;

import com.nimbusds.jose.jwk.JWKSet;
import com.nimbusds.jose.jwk.RSAKey;
import com.nimbusds.jose.jwk.source.ImmutableJWKSet;
import io.github.kingg22.api.vacunas.panama.modules.usuario.service.ITokenService;
import io.github.kingg22.api.vacunas.panama.service.UserDetailsServiceImpl;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.ProviderManager;
import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
import org.springframework.security.authentication.password.CompromisedPasswordChecker;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
import org.springframework.security.oauth2.jwt.JwtDecoder;
import org.springframework.security.oauth2.jwt.JwtEncoder;
import org.springframework.security.oauth2.jwt.JwtValidators;
import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
import org.springframework.security.oauth2.jwt.NimbusJwtEncoder;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
import org.springframework.security.oauth2.server.resource.web.authentication.BearerTokenAuthenticationFilter;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.security.web.authentication.password.HaveIBeenPwnedRestApiPasswordChecker;

@Configuration
@EnableWebSecurity
@EnableMethodSecurity(securedEnabled = true)
@RequiredArgsConstructor
public class SecurityConfig {
    @Value("${security.jwt.issuer}")
    private String issuer;

    private final RSAPrivateKey privateKey;
    private final RSAPublicKey publicKey;
    private final AccessDeniedHandler accessDeniedHandler;
    private final AuthenticationEntryPoint authenticationEntryPoint;

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http, ITokenService tokenService, JwtDecoder jwtDecoder)
            throws Exception {
        return http.csrf(csrf -> csrf.ignoringRequestMatchers("/vacunacion/**"))
                .headers(headers -> headers.contentSecurityPolicy(cspc -> cspc.policyDirectives(
                        "default-src 'none'; frame-ancestors 'none'; sandbox; media-src 'self'; object-src 'self';")))
                .authorizeHttpRequests(auth -> auth.requestMatchers("/vacunacion/v1/account/register")
                        .permitAll()
                        .requestMatchers("/vacunacion/v1/account/login")
                        .permitAll()
                        .requestMatchers("/vacunacion/v1/account/restore/**")
                        .permitAll()
                        .requestMatchers("/vacunacion/v1/public/**")
                        .permitAll()
                        .requestMatchers("/vacunacion/v1/bulk/**")
                        .permitAll()
                        .requestMatchers("/vacunacion/v1/pdf/**")
                        .permitAll()
                        .requestMatchers("/vacunacion/v1/patient/**")
                        .hasAnyAuthority("PACIENTE_READ")
                        .requestMatchers("/vacunacion/v1/vaccines/**")
                        .hasAnyRole("DOCTOR", "ENFERMERA")
                        .anyRequest()
                        .authenticated())
                .addFilterAfter(
                        new CustomJwtRefreshFilter(tokenService, jwtDecoder), BearerTokenAuthenticationFilter.class)
                .exceptionHandling(exception -> exception
                        .accessDeniedHandler(accessDeniedHandler)
                        .authenticationEntryPoint(authenticationEntryPoint))
                .oauth2ResourceServer(oauth2 -> oauth2.jwt(Customizer.withDefaults()))
                .sessionManagement(session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .build();
    }

    @Bean
    JwtAuthenticationConverter jwtAuthenticationConverter() {
        var grantedAuthoritiesConverter = new JwtGrantedAuthoritiesConverter();
        grantedAuthoritiesConverter.setAuthorityPrefix("");
        grantedAuthoritiesConverter.setAuthoritiesClaimName("scope");
        var converter = new JwtAuthenticationConverter();
        converter.setJwtGrantedAuthoritiesConverter(grantedAuthoritiesConverter);
        return converter;
    }

    @Bean
    JwtDecoder jwtDecoder(ITokenService tokenService) {
        var nimbusJwtDecoder = NimbusJwtDecoder.withPublicKey(this.publicKey)
                .signatureAlgorithm(SignatureAlgorithm.RS256)
                .build();
        nimbusJwtDecoder.setJwtValidator(JwtValidators.createDefaultWithIssuer(issuer));
        nimbusJwtDecoder.setJwtValidator(new CustomRedisJwtValidator(tokenService));
        return nimbusJwtDecoder;
    }

    @Bean
    JwtEncoder jwtEncoder() {
        var jwk = new RSAKey.Builder(this.publicKey).privateKey(this.privateKey).build();
        return new NimbusJwtEncoder(new ImmutableJWKSet<>(new JWKSet(jwk)));
    }

    @Bean
    BCryptPasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public CompromisedPasswordChecker compromisedPasswordChecker() {
        return new HaveIBeenPwnedRestApiPasswordChecker();
    }

    @Bean
    AuthenticationManager authenticationManager(
            UserDetailsServiceImpl userDetailsServiceImpl,
            PasswordEncoder passwordEncoder,
            CompromisedPasswordChecker compromisedPasswordChecker) {
        var daoAuthenticationProvider = new DaoAuthenticationProvider(passwordEncoder);
        daoAuthenticationProvider.setUserDetailsService(userDetailsServiceImpl);
        daoAuthenticationProvider.setCompromisedPasswordChecker(compromisedPasswordChecker);
        return new ProviderManager(daoAuthenticationProvider);
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/configuration/security/SecurityConfig.kt
@file:Suppress("ktlint:standard:no-empty-file")

package io.github.kingg22.api.vacunas.panama.configuration.security

/*
TODO all configuration need be change to quarkus or something
@Configuration
@EnableWebFluxSecurity
@EnableReactiveMethodSecurity
class SecurityConfig(
    @Value("\${security.jwt.issuer}") private val issuer: String,
    private val privateKey: RSAPrivateKey,
    private val publicKey: RSAPublicKey,
    private val accessDeniedHandler: ServerAccessDeniedHandler,
    private val authenticationEntryPoint: ServerAuthenticationEntryPoint,
) {
    @Bean
    fun securityFilterChain(
        http: ServerHttpSecurity,
        reactiveJwtDecoder: ReactiveJwtDecoder,
        reactiveJwtAuthenticationConverter: ReactiveJwtAuthenticationConverterAdapter,
        jwtRefreshFilter: CustomJwtRefreshFilter,
    ): SecurityWebFilterChain = http
        .csrf {
            it.requireCsrfProtectionMatcher(
                NegatedServerWebExchangeMatcher(PathPatternParserServerWebExchangeMatcher("/**/")),
            )
        }
        .headers {
            it.contentSecurityPolicy { cspc ->
                cspc.policyDirectives(
                    "default-src 'none'; frame-ancestors 'none'; sandbox; media-src 'self'; object-src 'self';",
                )
            }
        }
        .authorizeExchange {
            it.pathMatchers(
                "/account/register",
                "/account/login",
                "/account/restore/**/",
                "/public/**/",
                "/bulk/**/",
                "/direccion/provincias",
                "/direccion/distritos",
                "/sedes",
                "/vaccines",
                "/roles",
                "/roles/permisos",
            ).permitAll()
                .pathMatchers("/patient/**/", "/pdf/**/").hasAnyAuthority("PACIENTE_READ")
                .pathMatchers("/vaccines/**/").hasAnyRole("DOCTOR", "ENFERMERA")
                .anyExchange().authenticated()
        }
        .addFilterAfter(jwtRefreshFilter, SecurityWebFiltersOrder.AUTHENTICATION) // check this
        .exceptionHandling {
            it.accessDeniedHandler(accessDeniedHandler).authenticationEntryPoint(authenticationEntryPoint)
        }
        .oauth2ResourceServer {
            it.jwt { jwt ->
                jwt.jwtDecoder(reactiveJwtDecoder)
                jwt.jwtAuthenticationConverter(reactiveJwtAuthenticationConverter)
            }
        }
        .requestCache { it.disable() }
        .securityContextRepository(NoOpServerSecurityContextRepository.getInstance())
        .formLogin { it.disable() }
        .logout { it.disable() }
        .httpBasic { it.disable() }
        .build()

    @Bean
    fun reactiveJwtAuthenticationConverter(): ReactiveJwtAuthenticationConverterAdapter {
        val grantedAuthoritiesConverter = JwtGrantedAuthoritiesConverter()
        grantedAuthoritiesConverter.setAuthorityPrefix("")
        grantedAuthoritiesConverter.setAuthoritiesClaimName("scope")
        val converter = JwtAuthenticationConverter()
        converter.setJwtGrantedAuthoritiesConverter(grantedAuthoritiesConverter)
        return ReactiveJwtAuthenticationConverterAdapter(converter)
    }

    @Bean
    fun reactiveJwtDecoder(): ReactiveJwtDecoder {
        val nimbusJwtDecoder = NimbusReactiveJwtDecoder.withPublicKey(publicKey)
            .signatureAlgorithm(SignatureAlgorithm.RS256)
            .build()
        nimbusJwtDecoder.setJwtValidator(JwtValidators.createDefaultWithIssuer(issuer))
        return nimbusJwtDecoder
    }

    @Bean
    fun jwtEncoder(): JwtEncoder {
        val jwk =
            RSAKey.Builder(this.publicKey).privateKey(this.privateKey).build()
        return NimbusJwtEncoder(ImmutableJWKSet(JWKSet(jwk)))
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder = BCryptPasswordEncoder()

    @Bean
    fun reactiveCompromisedPasswordChecker(): ReactiveCompromisedPasswordChecker =
        HaveIBeenPwnedRestApiReactivePasswordChecker()

    @Bean
    fun reactiveAuthenticationManager(
        userDetailsServiceImpl: ReactiveUserDetailsService,
        passwordEncoder: PasswordEncoder,
        compromisedPasswordChecker: ReactiveCompromisedPasswordChecker,
    ): ReactiveAuthenticationManager {
        val provider = UserDetailsRepositoryReactiveAuthenticationManager(userDetailsServiceImpl)
        provider.setPasswordEncoder(passwordEncoder)
        provider.setCompromisedPasswordChecker(compromisedPasswordChecker)
        return provider
    }
}
*/
